### PR TITLE
Add support for multi-document YAML files

### PR DESCRIPTION
**Pull request by dan-cohn**
I discovered that pretty_format_yaml doesn't work with YAML files containing multiple "documents" in a single stream.  It always fails on these files because it's unable to parse them.  Fortunately, ruamel.yaml provides a simple way to load multiple documents from a file.  It also provides a way to dump them back out.  I took advantage of these functions so it now works on single- and multi-doc files.  New tests are included as well.

### CODE REVIEW

File: tests/pretty_format_yaml_test.py [hunk 1]
```diff
@@ -45,3 +47,16 @@ def test_pretty_format_yaml_autofix(tmpdir):
     # file was formatted (shouldn't trigger linter again)
     ret = pretty_format_yaml([srcfile.strpath])
     assert ret == 0
+
+
+def test_pretty_format_multidoc_yaml_autofix(tmpdir):
```
Comment (macisamuele) [inline tests/pretty_format_yaml_test.py:52]: Nit (not-blocking): The new test can be replaces with a fixture on `test_pretty_format_yaml_autofix` so we can avoid duplicating code.

Example
```python
@pytest.fixture('no_pretty_file_name', ['not-pretty-formatted.yaml', 'multi-doc-not-pretty-formatted.yaml'])
def test_pretty_format_yaml_autofix(tmpdir, no_pretty_file_name):
   ...
```

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py [hunk 2]
```diff
@@ -47,14 +47,17 @@ def pretty_format_yaml(argv=None):
...
+            if len(content) == 1:
+                if not isinstance(content[0], (list, dict)):
+                    # skip files containing primitive types (unstructured text)
+                    continue
 
-            pretty_content = StringIO()
-            yaml.dump(content, pretty_content)
+                yaml.dump(content[0], pretty_content)
+            else:
+                yaml.dump_all(content, pretty_content)
```
Comment (macisamuele) [inline language_formatters_pre_commit_hooks/pretty_format_yaml.py:60]: The check that excludes from pretty-fication the primitive types needs to be done for all the documents (more context in [here](https://github.com/macisamuele/language-formatters-pre-commit-hooks/pull/1#discussion_r370216799)).

I do believe that the current implementation has the potential of creating a new occurrence of #1 .

In order to work around this issue I think that we might pretty-format each individual document and then concat them. Something like the following
```python
pretty_docs = []
for ydoc in yaml.load_all(...):
    if isinstance(ydoc, (list, dict)):
        pretty_content = StringIO()
        yaml.dump(ydoc, pretty_content)
        pretty_docs.append(pretty_content.getvalue())
    else:
        # skip files containing primitive types (unstructured text)	
        pretty_docs.append(str(ydoc))

output = '\n---'.join(pretty_docs)
```
(disclaimer: I have not actually tried the code, so it could be very buggy)

File: test-data/pretty_format_yaml/multi-doc-not-pretty-formatted.yaml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+---
+root:
+    test:
+     -  1
```
Comment (macisamuele) [inline test-data/pretty_format_yaml/multi-doc-not-pretty-formatted.yaml:4]: I would make this doc very ugly, so the result will be more visible
```suggestion
root: {test: [1]}
```

File: test-data/pretty_format_yaml/multi-doc-pretty-formatted.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+root:
+  test:
+  - 1
+---
+second_root:
+  test:
+  - 2
```
Comment (macisamuele) [inline test-data/pretty_format_yaml/multi-doc-pretty-formatted.yaml:7]: In case of multi-doc YAML files I think that having the explicit document start (`---`) would be beneficial.

File: tests/pretty_format_yaml_test.py (hunk 1 above)
Comment (dan-cohn) [reply tests/pretty_format_yaml_test.py:52]: Good suggestion.  I'm happy to update.  I have very little experience with pytest, so it was mostly ignorance on my part.

File: test-data/pretty_format_yaml/multi-doc-pretty-formatted.yaml (hunk 4 above)
Comment (dan-cohn) [reply test-data/pretty_format_yaml/multi-doc-pretty-formatted.yaml:7]: I thought about this, but I was on the fence because the library doesn't include it with `dump_all()`.  Also, I've seen files with and without it.  What do you think about making this configurable?  Something like "--omit-document-start" to leave it out.

Comment (macisamuele) [reply test-data/pretty_format_yaml/multi-doc-pretty-formatted.yaml:7]: I would say that adding configurations would not really be helpful and anyway this is not a blocking issue.

By the way, https://github.com/macisamuele/language-formatters-pre-commit-hooks/pull/3#discussion_r418655126 requires some changes that most probably will move you away from using `dump_all` ;)

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py (hunk 2 above)
Comment (macisamuele) [reply language_formatters_pre_commit_hooks/pretty_format_yaml.py:60]: NOTE: You might try to visualise the mentioned issue by running the following code
```python
from io import StringIO
from ruamel.yaml import YAML
yaml_content  = "---\n{}\n---\na: 1".format('a ' * 100)

fixed_content = StringIO()
YAML().dump_all(documents=list(YAML().load_all(yaml_content)), stream=fixed_content)

print("Initial content:")
print(yaml_content)
print("Fixed content:")
print(fixed_content.getvalue())

assert YAML().load_all(yaml_content) == YAML().load_all(fixed_content.getvalue())
```
Output
```
Initial content:
---
a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a
---
a: 1
Fixed content:
a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a
a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a
a a a a a a a a a a a a a a a a a a
---
a: 1

Traceback (most recent call last):
  File "a.py", line 13, in <module>
    assert YAML().load_all(yaml_content) == YAML().load_all(fixed_content.getvalue())
AssertionError
```
You can see how the first document is different (and as it is a string it is actually a different string

FYI I've noticed that this phenomena does actually happen also if the string is used as value of a mapping (#4)

File: test-data/pretty_format_yaml/multi-doc-not-pretty-formatted.yaml (hunk 3 above)
Comment (dan-cohn) [reply test-data/pretty_format_yaml/multi-doc-not-pretty-formatted.yaml:4]: I can make it more ugly, but your suggestion is already pretty-formatted according to the tool.

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py (hunk 2 above)
Comment (dan-cohn) [reply language_formatters_pre_commit_hooks/pretty_format_yaml.py:60]: I thought about this, but I'm not sure it's legit to have a real YAML file that contains non-YAML (primitive) documents.  Can you point me to a real-world example of this?

File: test-data/pretty_format_yaml/multi-doc-not-pretty-formatted.yaml (hunk 3 above)
Comment (macisamuele) [reply test-data/pretty_format_yaml/multi-doc-not-pretty-formatted.yaml:4]: That's surprising ... but yeah. Let's ignore this comment.
I'm marking this as resolved

File: tests/pretty_format_yaml_test.py (hunk 1 above)
Comment (macisamuele) [reply tests/pretty_format_yaml_test.py]: Don't be harsh on yourself ... every day is a good day to learn new things

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py (hunk 2 above)
Comment (macisamuele) [reply language_formatters_pre_commit_hooks/pretty_format_yaml.py]: I don't really have an example here, personally I use YAML mostly for swagger specs or writing configurations (so always mappings and single docs).

The specifications do actually allow to have primitive types at the root level and so the following file should be considered valid
```yaml
---
null
---
this is a string
---
1
---
true
---
1.0
---
- first item
- second item
---
key1: value 1
key2: value 2
```

Considering this, I think that we should handle such case as well (especially considering that the effort to do so is not huge)

Comment (dan-cohn) [reply language_formatters_pre_commit_hooks/pretty_format_yaml.py]: OK, fair enough. Shouldn't be a problem to add it and maybe include a test or two as well.

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py [hunk 5]
```diff
@@ -42,27 +43,38 @@ def pretty_format_yaml(argv=None):
...
+        separator_pattern = '^--- ?.*\\n'
+        original_docs = re.split(separator_pattern, string_content, flags=re.MULTILINE)
+        pretty_docs = []
 
-            pretty_content = StringIO()
-            yaml.dump(content, pretty_content)
-
-            if string_content != pretty_content.getvalue():
+        try:
+            for doc in original_docs:
```
Comment (macisamuele) [inline language_formatters_pre_commit_hooks/pretty_format_yaml.py:57]: Please use `load_all` from `ruamel.yaml` (as proposed in https://github.com/macisamuele/language-formatters-pre-commit-hooks/pull/3/#discussion_r418655126)

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py [hunk 6]
```diff
@@ -42,31 +43,38 @@ def pretty_format_yaml(argv=None):
...
 
-        try:
-            content = list(yaml.load_all(string_content))
-            pretty_content = StringIO()
-
-            if len(content) == 1:
-                if not isinstance(content[0], (list, dict)):
-                    # skip files containing primitive types (unstructured text)
-                    continue
+        separator_pattern = '^--- ?.*\\n'
```
[the remark this answers is not in the document]
Comment (dan-cohn) [reply language_formatters_pre_commit_hooks/pretty_format_yaml.py:52]: You're right about the standard. I decided to be more flexible because of examples like the one [here](https://yaml.org/start.html). Nonetheless, it makes no difference to me. I'm happy to change it.

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py [hunk 7]
```diff
@@ -42,27 +43,38 @@ def pretty_format_yaml(argv=None):
...
             string_content = ''.join(f.readlines())
 
-        try:
-            content = yaml.load(string_content)
-
-            if not isinstance(content, (list, dict)):
-                # skip files containing primitive types (unstructured text)
-                continue
+        separator_pattern = '^--- ?.*\\n'
+        original_docs = re.split(separator_pattern, string_content, flags=re.MULTILINE)
```
[the remark this answers is not in the document]
Comment (dan-cohn) [reply language_formatters_pre_commit_hooks/pretty_format_yaml.py:53]: I agree in principle, but I tried using `load_all` and it creates a problem for primitive documents. It removes all the newlines. This causes even a file with just one document in it to get messed up. The Ansible Vault file is a good example.

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py (hunk 5 above)
Comment (dan-cohn) [reply language_formatters_pre_commit_hooks/pretty_format_yaml.py:57]: As I mentioned, it isn't feasible because of what it does with newlines. My current approach seems to work fine. It uses load for everything except detecting the doc separators.

**Comment by dan-cohn** [comment]
@macisamuele Please have a look at my latest code. Is this what you had in mind?

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py [hunk 8]
```diff
@@ -14,6 +15,28 @@
 from six import text_type
 
 
+def process_single_document(document, yaml):
```
Comment (macisamuele) [inline language_formatters_pre_commit_hooks/pretty_format_yaml.py:18]: I would make the function private (I don't expect that anyone uses this library as dependency of their code, but its easy to address it at this stage)

```suggestion
def _process_single_document(document, yaml):
```

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py [hunk 9]
```diff
@@ -14,6 +15,28 @@
...
+    """Pretty format one document containing YAML or primitive (non-YAML) text.
+    Call this function for each doc within a multi-document .yaml file.
+
+    Args:
+        document (str): Original document content.
+        yaml: YAML library instance.
+
+    Returns:
+        Pretty-formatted content (str).
+    """
```
Comment (macisamuele) [inline language_formatters_pre_commit_hooks/pretty_format_yaml.py:28]: Small rewording, mostly needed to generalise the documentation as the task done by the method is pretty general

```suggestion
    """Pretty format one YAML document.
    
    This is needed in order to prevent `ruamel.yaml` to interfere with documents that have primitive types on the document root.
     For more context check https://github.com/macisamuele/language-formatters-pre-commit-hooks/pull/1

    Args:
        document (str): Original document content.
        yaml: YAML library instance.
    Returns:
        Pretty-formatted content (str).
    """
```

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py [hunk 10]
```diff
@@ -14,6 +15,28 @@
...
+    content = yaml.load(document)
+    if isinstance(content, (list, dict)):
+        pretty_output = StringIO()
+        yaml.dump(content, pretty_output)
+        return pretty_output.getvalue()
+    elif document:
+        # do not disturb primitive content (unstructured text)
+        return str(document)
+    return None
```
Comment (macisamuele) [inline language_formatters_pre_commit_hooks/pretty_format_yaml.py:37]: I would rather return document directly as it is already of type `str` (docstring).
Additionally returning the document if evaluate to `True` means that an empty document (empty string) would be removed from the output documents (according to the around line 80)

```suggestion
    # do not disturb primitive content (ie. unstructured text)
    return document
```

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py [hunk 11]
```diff
@@ -42,27 +65,38 @@ def pretty_format_yaml(argv=None):
...
         with open(yaml_file) as f:
             string_content = ''.join(f.readlines())
 
-        try:
-            content = yaml.load(string_content)
+        # Split multi-document file into individual documents
+        #
+        # Not using yaml.load_all() because it reformats primitive (non-YAML) content. It removes
+        # newline characters.
+        separator_pattern = '^---\\s*\\n'
```
Comment (macisamuele) [inline language_formatters_pre_commit_hooks/pretty_format_yaml.py:78]: nit.
```suggestion
        separator_pattern = r'^---\s*\n'
```

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py [hunk 12]
```diff
@@ -42,27 +65,38 @@ def pretty_format_yaml(argv=None):
...
-            if not isinstance(content, (list, dict)):
-                # skip files containing primitive types (unstructured text)
-                continue
+        pretty_docs = []
+
+        try:
+            for doc in original_docs:
+                content = process_single_document(doc, yaml)
+                if content is not None:
+                    pretty_docs.append(content)
```
Comment (macisamuele) [inline language_formatters_pre_commit_hooks/pretty_format_yaml.py:87]:
```suggestion
                pretty_docs.append(_ process_single_document(doc, yaml))
```

File: test-data/pretty_format_yaml/multi-doc-pretty-formatted.yaml [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+  - hi there
+---
+second_root:
+  test:
+  - 1
+  - 2
+---
+This is a primitive file.
+
+Last line of file.
```
Comment (macisamuele) [inline test-data/pretty_format_yaml/multi-doc-pretty-formatted.yaml:13]: I would put two consecutive text lines as well.
Suggesting this because on https://repl.it/@macisamuele/language-formatters-pre-commit-hooks-PR3 I've noticed that
```
aaa
bbb

ccc
```
would be read as `'aaa bbb\nccc'`

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py (hunk 10 above)
Comment (dan-cohn) [reply language_formatters_pre_commit_hooks/pretty_format_yaml.py:38]: I did this deliberately because otherwise it inserts blank documents, resulting in things like this:
```yaml
---
---
```
At this point, I would rather merge what I have and let you make any functional changes you like. I don't have time to test all the variations.

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py (hunk 11 above)
Comment (dan-cohn) [reply language_formatters_pre_commit_hooks/pretty_format_yaml.py:79]: I thought I tried that originally and it didn't work. But it seems to be fine.

File: test-data/pretty_format_yaml/multi-doc-pretty-formatted.yaml (hunk 13 above)
Comment (dan-cohn) [reply test-data/pretty_format_yaml/multi-doc-pretty-formatted.yaml:13]: Sure, but it works fine. The reason you had a problem is that you used `load_all()`.

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py (hunk 12 above)
Comment (dan-cohn) [reply language_formatters_pre_commit_hooks/pretty_format_yaml.py:88]: I confirmed that this fails my tests. Not making this change.

**Comment by dan-cohn** [comment]
I've done what I can.  I think the code is quite solid as this point.  As I mentioned in my earlier replies, I would rather not make a functional change by including empty documents.  In my testing, this produced undesirable results.  If you feel strongly about it, I'll let you make the change after merging the PR.  I would caution you to test it thoroughly including edge cases such as files containing only primitive data, files containing a single YAML document, files that end with ---, and files that begin with --- but only contain one YAML doc.

File: language_formatters_pre_commit_hooks/pretty_format_yaml.py (hunk 10 above)
Comment (macisamuele) [reply language_formatters_pre_commit_hooks/pretty_format_yaml.py:38]: ++ Created #5 to track this.
